src/hotspot/cpu/x86/stubGenerator_x86_64_poly.cpp

```diff
@@ -0,0 +1,1066 @@
+  __ vpsllq(D0, R2, 2, Assembler::AVX_512bit);
+  __ vpaddq(R2P, R2, D0, Assembler::AVX_512bit); // R2' (R2*5)
+  __ vpsllq(R1P, R1P, 2, Assembler::AVX_512bit); // 4*5*R
+  __ vpsllq(R2P, R2P, 2, Assembler::AVX_512bit);
+
+  poly1305_multiply8_avx512(A0, A1, A2,            // MSG/ACC 16 blocks
+                            B0, B1, B2, B3, B4,    // R^16-R^9, R1P, R2P
+                            T0, T1, T2, T3, T4, T5, TMP, rscratch);
+  poly1305_multiply8_avx512(A3, A4, A5,            // MSG/ACC 16 blocks
+                            R0, R1, R2, R1P, R2P,  // R^8-R, R1P, R2P
+                            T0, T1, T2, T3, T4, T5, TMP, rscratch);
+
+  // Add all blocks (horizontally)
+  // 16->8 blocks
+  __ vpaddq(A0, A0, A3, Assembler::AVX_512bit);
+  __ vpaddq(A1, A1, A4, Assembler::AVX_512bit);
+  __ vpaddq(A2, A2, A5, Assembler::AVX_512bit);
+
+  // 8 -> 4 blocks
+  __ vextracti64x4(D0, A0, 1);
+  __ vextracti64x4(D1, A1, 1);
+  __ vextracti64x4(D2, A2, 1);
+  __ vpaddq(A0, A0, D0, Assembler::AVX_256bit);
+  __ vpaddq(A1, A1, D1, Assembler::AVX_256bit);
+  __ vpaddq(A2, A2, D2, Assembler::AVX_256bit);
+
+  // 4 -> 2 blocks
+  __ vextracti32x4(D0, A0, 1);
+  __ vextracti32x4(D1, A1, 1);
+  __ vextracti32x4(D2, A2, 1);
+  __ vpaddq(A0, A0, D0, Assembler::AVX_128bit);
+  __ vpaddq(A1, A1, D1, Assembler::AVX_128bit);
+  __ vpaddq(A2, A2, D2, Assembler::AVX_128bit);
+
+  // 2 -> 1 blocks
+  __ vpsrldq(D0, A0, 8, Assembler::AVX_128bit);
+  __ vpsrldq(D1, A1, 8, Assembler::AVX_128bit);
+  __ vpsrldq(D2, A2, 8, Assembler::AVX_128bit);
+
+  // Finish folding and clear second qword
+  __ mov64(t0, 0xfd);
+  __ kmovql(k1, t0);
+  __ evpaddq(A0, k1, A0, D0, false, Assembler::AVX_512bit);
+  __ evpaddq(A1, k1, A1, D1, false, Assembler::AVX_512bit);
+  __ evpaddq(A2, k1, A2, D2, false, Assembler::AVX_512bit);
+
+  // Carry propagation
+  __ vpsrlq(D0, A0, 44, Assembler::AVX_512bit);
+  __ evpandq(A0, A0, ExternalAddress(poly1305_mask44()), Assembler::AVX_512bit, rscratch); // Clear top 20 bits
+  __ vpaddq(A1, A1, D0, Assembler::AVX_512bit);
+  __ vpsrlq(D0, A1, 44, Assembler::AVX_512bit);
+  __ evpandq(A1, A1, ExternalAddress(poly1305_mask44()), Assembler::AVX_512bit, rscratch); // Clear top 20 bits
+  __ vpaddq(A2, A2, D0, Assembler::AVX_512bit);
+  __ vpsrlq(D0, A2, 42, Assembler::AVX_512bit);
+  __ evpandq(A2, A2, ExternalAddress(poly1305_mask42()), Assembler::AVX_512bit, rscratch); // Clear top 22 bits
+  __ vpsllq(D1, D0, 2, Assembler::AVX_512bit);
+  __ vpaddq(D0, D0, D1, Assembler::AVX_512bit);
+  __ vpaddq(A0, A0, D0, Assembler::AVX_512bit);
+
+  // Put together A (accumulator)
+  __ movq(a0, A0);
+
+  __ movq(t0, A1);
+  __ movq(t1, t0);
+  __ shlq(t1, 44);
+  __ shrq(t0, 20);
+
+  __ movq(a2, A2);
+  __ movq(a1, a2);
+  __ shlq(a1, 24);
+  __ shrq(a2, 40);
+
+  __ addq(a0, t1);
+  __ adcq(a1, t0);
+  __ adcq(a2, 0);
+
+  // Cleanup
+  __ vpxorq(xmm0, xmm0, xmm0, Assembler::AVX_512bit);
+  __ vpxorq(xmm1, xmm1, xmm1, Assembler::AVX_512bit);
```

Comment:
You could use T0, T1 in place of xmm0, xmm1 here.

Reply:
Or simply switch to `vzeroall` for `xmm0` - `xmm15`.

Reply:
ah.. I remember thinking about doing that.. `vzeroall` isnt encoded yet and I figured since I already have to do the xmm16-29, might as well do them all.. should I add that instruction too?

Reply:
Yes, please. And for the upper half of register file, just code it as a loop over register range:
```
for (int rxmm_num = 16; rxmm_num < 30; rxmm_num++) {
  XMMRegister rxmm = as_XMMRegister(rxmm_num);
  __ vpxorq(rxmm, rxmm, rxmm, Assembler::AVX_512bit);
}
```
or even
```
// Zeroes zmm16-zmm31. 
for (XMMRegister rxmm = xmm16; rxmm->is_valid(); rxmm = rxmm->successor()) {
  __ vpxorq(rxmm, rxmm, rxmm, Assembler::AVX_512bit);
}
```

Reply:
Will do.. ("loop" erm.. wow.. "duh, this isn't assembler!") Thanks!!

Reply:
done